### ISSUE TITLE
QE for HTML input doesn't work

### ISSUE THREAD

**Issue by abhi-agg** [browsermt/bergamot-translator#355]
A quick experiment with wasm test page for html text translation shows weird byte ranges for words.

An example:

```
Sentence: "How are  <b>you</b>  doing?"
Translated sentence: "Wie geht <b>es dir</b>?"

"words":["Wie", "geht", "<b", "es d"]
"wordByteRanges":[{"begin":0,"end":3}, {"begin":4,"end":8}, {"begin":9,"end":11}, {"begin":12,"end":16}]

```

Looks like, QE ignores tags completely (treating them as if they are non-existent) in the translated sentences and compute the byte ranges of the words in the translated sentences. Is it something that needs to be fixed or am I doing something wrong at my end?

Attaching the image for detailed results.

<img width="1911" alt="Screenshot 2022-02-17 at 14 32 35" src="https://user-images.githubusercontent.com/66322306/154493114-2f9c8450-e05f-492b-9500-d06de7ba5b53.png">

cc @kpu @jerinphilip @abarbosa94 @felipesantosk @mfomicheva

**Comment by abhi-agg** [comment]
> Looks like, QE ignores tags completely (treating them as if they are non-existent) in the translated sentences and compute the byte ranges of the words in the translated sentences.

Because I translated `How are you doing?` after removing all the tags from html sentence and I get the same byte ranges for both cases.

Attached image for reference.
<img width="1911" alt="Screenshot 2022-02-17 at 14 49 19" src="https://user-images.githubusercontent.com/66322306/154495202-5baf35df-7c9f-4220-a828-c54a011f1d9d.png">

**Comment by jelmervdl** [comment]
Ah, I totally missed that. I was under the impression that QualityEstimation scores uses those same offsets that are stored in the `AnnotatedText` class that HTML updates. It does… but it aggregates them into words and then stores its own byte offsets all before the HTML is inserted into the translation.

I'll see whether I can make the `QualityEstimator` class work on token offsets instead. Then it automatically stays up to date with the byte offset updates that HTML makes. (And it would also benefit from changes like #298.)

**Comment by jerinphilip** [comment]
QE is slated to [be implemented in W5](https://github.com/mozilla/firefox-translations/issues/26#event-6039801870) (March 7-11). HTML is [expected to be implemented in W3](https://github.com/mozilla/firefox-translations/issues/52#event-6039402149) (February 21-25), which is next week. I guess it's reasonable to keep this soft at W4 here and prioritize based on how HTML completes or not by W3.

**Comment by abhi-agg** [comment]
> QE is slated to [be implemented in W5](https://github.com/mozilla/firefox-translations/issues/26#event-6039801870) (March 7-11). HTML is [expected to be implemented in W3](https://github.com/mozilla/firefox-translations/issues/52#event-6039402149) (February 21-25), which is next week. I guess it's reasonable to keep this soft at W4 here and prioritize based on how HTML completes or not by W3.

The whole idea of filing this issue was based on @kpu's suggestion (btw @kpu thanks a lot for this suggestion 👍🏾) in last plenary to start integrating and find bugs quickly so that technical support from involving partners could be solicited asap.

QE feature is definitely going to be a part of the final delivery and so is the HTML. In my opinion, this fix doesn't have to wait till W4. We need to test QE again after this bug is fixed and that might resurface some new issues which further might take some time to fix.

All of this goes in general for anything that is becoming a part of the final delivery.
cc @andrenatal @lonnen

**Comment by kpu** [comment]
@jelmervdl is working on it.

**Comment by abhi-agg** [comment]
> My idea was to remove the `subwordToWord` call here:
> 
> https://github.com/browsermt/bergamot-translator/blob/2844cedb0dc8b67df5b656e61b69e194a6658501/src/translator/quality_estimator.cpp#L30
> 
> That call just turns the token ranges that QualityEstimator already uses internally into byte ranges:
> 
> https://github.com/browsermt/bergamot-translator/blob/2844cedb0dc8b67df5b656e61b69e194a6658501/src/translator/quality_estimator.cpp#L270-L285
> 
> That step is something a client can easily do itself, it has access to the AnnotatedText.

@jelmervdl Thanks for working on it. I would add one thing. If I remember correctly, the translation models (and perhaps  supervised QE models as well) provide scores of the "subwords" and then some post-processing is done to convert the "subword" scores to "word" scores. We discussed in the beginning with the QE team and @jerinphilip that it is better for the client to receive the quality scores at "word" level so that client doesn't need to do this post-processing. It is more efficient to do it in the engine as it has greater control and access to more information, plus avoiding replicating the same logic for every client. The client was kind of made oblivious to the concept of "subword" when the html translation came inside the engine.

Open for discussion here if keeping it the same way for html text poses a technical challenge 👍🏾

**Comment by kpu** [comment]
@abhi-agg @andrenatal Please take a look at #357 and #358.  
#357: Byte spans expanded to cover HTML removal/insertion.  
#358: explicit HTML tags inserted with QE scores as an attribute.  

Of course QE words and HTML tags are unsynchronized i.e. a QE word can span over an HTML open or closing tag.

**Comment by jelmervdl** [comment]
> We discussed in the beginning with the QE team and @jerinphilip that it is better for the client to receive the quality scores at "word" level so that client doesn't need to do this post-processing.

I agree with the reasoning here. In both #357 I effectively moved the `subwordToWord` call to the WASM bindings part. In Javascript you'll still receive scores per "word" (where the byterange for that word can include some HTML tags as well, but never more text)

In #358 you get this behaviour for free since I assign the same `<font x-bergamot-word-score="">` tag to all subwords in a word, and the HTML insertion part is smart enough to then group all subwords into the same font tag.

**Comment by kpu** [comment]
1. QE has only delivered three models and they're all out of English, so there is no bidirectional QE.  Which is said because QE might have been useful in that context. However, text/plain translation remains a requirement.  

2. There is a mockup.  See https://github.com/browsermt/coordination/blob/master/docs/D1.3-Bergamot_User_interface_with_quality_estimation.pdf in the private repo.

**Comment by jerinphilip** [comment]
> There is a mockup. See [https://github.com/browsermt/coordination/blob/master/docs/D1.3-Bergamot_User_interface_with_quality_estimation.pdf](https://github.com/browsermt/coordination/blob/master/docs/D1.3-Bergamot_User_interface_with_quality_estimation.pdf?rgh-link-date=2022-02-18T11%3A50%3A31Z) in the private repo.
I found the following underline based proposal, and a few variants in Section 5 relating to QE.

<img width="361" alt="image" src="https://user-images.githubusercontent.com/727292/154679245-e4ac3c70-8243-4a24-9e02-beba5bae4d92.png">

From meetings I think it's important to discuss the following. I think @jelmer's render might have to do the thresholding (in C++) that was earlier meant for @abhi-agg to do to get the probability values to classes (Section 5: Major/Critical, Major/Minor/Critical, Poor/Ok). Right now we're exporting real values as data. I'm not sure if the decision rule can be applied in CSS (I've lost track of how capable it has become), @jelmervdl's currently using continuous colouring.

**Comment by jelmervdl** [comment]
All mock-ups seem to assume plain text (e.g. a form input), not HTML. In that case, what is currently in main will be sufficient. #357 won't hurt, but the offsets weren't broken to begin with in this scenario.

To clarify a bit further: #357 just fixes the byte offsets in case the input is marked as HTML. If the input was text, there was no issue. So for form translation (which I assume are always plain text unless you're going to attempt auto translation in `contenteditable` elements…) it is not necessary.

#358 adds the quality scores as HTML to HTML output, but this only works if the input was HTML to begin with. Of course, it could also be used with plain text if you'd encode all entities in the text before sending it to the translator (and mark it as HTML). The output can easily be rendered like the mock-ups. Thresholding can easily be done in the extension using Javascript or CSS. The added tags that don't meet the threshold won't affect rendering. You might need to be a bit creative with horizontal padding to fill in the gaps between consecutive words that do meet the threshold though. The way I insert the tags, it will leave the spaces that are not part of the word outside any of the score tags. But that's not insurmountable.

**Comment by kpu** [comment]
I think we should get #357 in to close this issue as that addresses the original post.  @abhi-agg can mull the HTML offer.

**Comment by kpu** [comment]
Written to @abhi-agg and @andrenatal seeking comment on #357 and #358 and am coming to the notion #358 is a better option.  Blocked awaiting their feedback.

**Comment by abhi-agg** [comment]
I have few queries here:

1. Both https://github.com/browsermt/bergamot-translator/pull/357 and https://github.com/browsermt/bergamot-translator/pull/358 solve the same issue. i.e. returning QE scores for HTML text translation and **not the plain text translation**. 
2. For plain text translation, QE scores will continue to be returned as before (i.e. byte ranges representing the "words" and their corresponding scores). This means JS side will format the text for QE by processing these byte ranges and scores.

**Comment by kpu** [comment]
@abhi-agg Plain text remains as it was already working.  However, what I would recommend as the easiest thing for you is to follow the suggestion in: https://github.com/browsermt/bergamot-translator/issues/355#issuecomment-1044425813 specifically:

HTML: pick #358 and you get the tags back.  
text/plain: you **encode** it then submit as HTML.  Render the HTML.  

Then you don't have to handle byte ranges in either case.

**Comment by abhi-agg** [comment]
> HTML: pick #358 and you get the tags back.
> text/plain: you **encode** it then submit as HTML. Render the HTML (Note it will be a snippet of HTML, not a full page.)

Using this approach means always setting [HTML](https://github.com/browsermt/bergamot-translator/blob/main/src/translator/response_options.h#L22) flag to `true`. Right? And in that case wouldn't the engine crash for sentences like `Hello &lt; world`?

**Comment by kpu** [comment]
> > HTML: pick #358 and you get the tags back.
> > text/plain: you **encode** it then submit as HTML. Render the HTML (Note it will be a snippet of HTML, not a full page.)
> 
> Using this approach means always setting [HTML](https://github.com/browsermt/bergamot-translator/blob/main/src/translator/response_options.h#L22) flag to `true`. Right? And in that case wouldn't the engine crash for sentences like `Hello &lt; world`?

> > HTML: pick #358 and you get the tags back.
> > text/plain: you **encode** it then submit as HTML. Render the HTML (Note it will be a snippet of HTML, not a full page.)
> 
> Using this approach means always setting [ResponseOptions::HTML](https://github.com/browsermt/bergamot-translator/blob/main/src/translator/response_options.h#L22) flag to `true`. Right? And in that case wouldn't the engine crash for sentences like `Hello &lt; world` (e.g this [issue](https://github.com/mozilla/firefox-translations/issues/51)).

`Hello &lt; world` is a valid HTML snippet.  `Hello < world` is not.  That's what **encode** refers to.  

I suppose one could always have HTML mode on then implement text/plain translation by encode -> pass as HTML -> decode.  That would be less efficient though.  

For QE, to render the colors, you are presumably doing HTML rendering in all cases anyway, hence the suggestion of having us generate the HTML output even if the input is (escaped) text/plain.

**Comment by jelmervdl** [comment]
Little Javascript example of how to encode plain text to HTML for when you want to submit plain text to the engine but want HTML as output:

```js
function encodePlainTextAsHTML(text) {
  const div = document.createElement('div');
  div.appendChild(document.createTextNode(text));
  return div.innerHTML;
}
```

And then also set responseOptions.html = true.

**Comment by jerinphilip** [comment]
> All mock-ups seem to assume plain text (e.g. a form input), not HTML.

If plaintext (textarea, document) is being translated, I don't think there's a way to annotate it with reds or colors (https://stackoverflow.com/a/12831155/4565794) without converting to HTML first. 

Suggest taking a look at [Grammarly](https://app.grammarly.com/) for an example. It's essentially the same problem, highlight poor confidence tokens and leave the others be. The textarea or input field is kept hidden, a `contentEditable` div takes over. The rich text annotations go on the div which support HTML. The textarea or form field can be populated with just the text content.

**Comment by jelmervdl** [comment]
@jerinphilip I assume that people type in the translation input field, and the quality is shown in a separate translation output field (which isn't editable when the UI is in this state…). Only once the translation is accepted you'd need to convert the translation output field to plain text to be inserted in the actual form field.

**Comment by abhi-agg** [comment]
> `Hello &lt; world` is a valid HTML snippet. `Hello < world` is not. That's what **encode** refers to.

Oh. Thanks for correcting 👍🏾 

> Little Javascript example of how to encode plain text to HTML for when you want to submit plain text to the engine but want HTML as output:
> 
> ```js
> function encodePlainTextAsHTML(text) {
>   const div = document.createElement('div');
>   div.appendChild(document.createTextNode(text));
>   return div.innerHTML;
> }
> ```
> And then also set responseOptions.html = true.

@jelmervdl With `responseOptions.html = true`, the sentence `Hello < world` if passed as it is will break the engine but the encoded version of it will not. Right?

**Comment by jelmervdl** [comment]
@abhi-agg  Yep! With `responseOptions.html = true`:  
Okay: `Hello &lt; world`
Not okay: `Hello < world`

**Comment by abhi-agg** [comment]
@jelmervdl Thanks for clarifying. Please correct me if I am wrong but with the approach https://github.com/browsermt/bergamot-translator/pull/358#issue-1142083849, extension will still have to parse the html translation result to show color-coded words based on thresholds.

**Comment by kpu** [comment]
@abhi-agg How exactly are you showing the color-coded words?  Wouldn't the natural way to do that be rendering the HTML with some CSS?

**Comment by jelmervdl** [comment]
I've [added an example](https://github.com/browsermt/bergamot-translator/pull/358/commits/922c2d1b2520a057ea7edfccc36df2cb90e3e0da) of how the output of #358 can be used to the demo page in that branch. That example treats the input field as plain text, escapes it, and renders the translation output as HTML. Then with a little bit of Javascript for thresholding and CSS I render sentence and word level quality indicators.

![image](https://user-images.githubusercontent.com/198639/155035311-83dce9da-4410-4dcb-8203-3e42cfe0be43.png)

 (Based on thresholds that have no meaning, I just wanted some thresholds that are met often enough for screenshot purposes.)
